File: projects/companion-light/v17.1/android/app/src/main/java/org/memoryalpha/companion/adapters/FrontierHttp.kt
// CC0-1.0
// BYOK frontier clients. Keys come from EncryptedSharedPreferences; the core never persists them.
package org.memoryalpha.companion.adapters

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import org.memoryalpha.companion.core.FrontierClient
import java.net.HttpURLConnection
import java.net.URL

private fun post(url: String, headers: Map<String, String>, body: JSONObject): JSONObject {
    val c = URL(url).openConnection() as HttpURLConnection
    c.requestMethod = "POST"; c.doOutput = true; c.setRequestProperty("Content-Type", "application/json")
    headers.forEach { (k, v) -> c.setRequestProperty(k, v) }
    c.outputStream.use { it.write(body.toString().toByteArray()) }
    val text = (if (c.responseCode < 300) c.inputStream else c.errorStream).bufferedReader().readText()
    if (c.responseCode >= 300) throw RuntimeException("HTTP ${c.responseCode}: $text")
    return JSONObject(text)
}

class AnthropicClient(private val key: String, private val model: String = "claude-sonnet-4-6") : FrontierClient {
    override val name = "claude"
    override suspend fun complete(system: String, user: String): String = withContext(Dispatchers.IO) {
        val j = post("https://api.anthropic.com/v1/messages", mapOf("x-api-key" to key, "anthropic-version" to "2023-06-01"),
            JSONObject().put("model", model).put("max_tokens", 1024).put("system", system)
                .put("messages", JSONArray().put(JSONObject().put("role", "user").put("content", user))))
        val parts = j.getJSONArray("content"); (0 until parts.length()).joinToString("") { parts.getJSONObject(it).optString("text") }
    }
}

class OpenAIClient(private val key: String, private val model: String = "gpt-5") : FrontierClient {
    override val name = "gpt"
    override suspend fun complete(system: String, user: String): String = withContext(Dispatchers.IO) {
        val j = post("https://api.openai.com/v1/chat/completions", mapOf("Authorization" to "Bearer $key"),
            JSONObject().put("model", model).put("messages", JSONArray()
                .put(JSONObject().put("role", "system").put("content", system)).put(JSONObject().put("role", "user").put("content", user))))
        j.getJSONArray("choices").getJSONObject(0).getJSONObject("message").getString("content")
    }
}

class GeminiClient(private val key: String, private val model: String = "gemini-2.5-pro") : FrontierClient {
    override val name = "gemini"
    override suspend fun complete(system: String, user: String): String = withContext(Dispatchers.IO) {
        val j = post("https://generativelanguage.googleapis.com/v1beta/models/$model:generateContent?key=$key", emptyMap(),
            JSONObject().put("system_instruction", JSONObject().put("parts", JSONArray().put(JSONObject().put("text", system))))
                .put("contents", JSONArray().put(JSONObject().put("role", "user").put("parts", JSONArray().put(JSONObject().put("text", user))))))
        val parts = j.getJSONArray("candidates").getJSONObject(0).getJSONObject("content").getJSONArray("parts")
        (0 until parts.length()).joinToString("") { parts.getJSONObject(it).optString("text") }
    }
}

File: projects/companion-light/v17.1/ios/CompanionLight/Sources/CompanionCore/Adapters/FrontierHTTP.swift
// CC0-1.0
// BYOK frontier clients. Keys live in the Keychain; passed in here, never persisted by the core.
import Foundation

public struct AnthropicClient: FrontierClient {
    public let name = "claude"; let key: String; let model: String
    public init(key: String, model: String = "claude-sonnet-4-6") { self.key = key; self.model = model }
    public func complete(system: String, user: String) async throws -> String {
        var req = URLRequest(url: URL(string: "https://api.anthropic.com/v1/messages")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.setValue(key, forHTTPHeaderField: "x-api-key")
        req.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        req.httpBody = try JSONSerialization.data(withJSONObject: ["model": model, "max_tokens": 1024, "system": system, "messages": [["role": "user", "content": user]]])
        let (data, resp) = try await URLSession.shared.data(for: req)
        try Self.check(resp, data)
        let j = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        return ((j["content"] as? [[String: Any]]) ?? []).compactMap { $0["text"] as? String }.joined()
    }
    static func check(_ r: URLResponse, _ d: Data) throws {
        if let h = r as? HTTPURLResponse, h.statusCode >= 300 { throw RuntimeError.http(h.statusCode, String(data: d, encoding: .utf8) ?? "") }
    }
}

public struct OpenAIClient: FrontierClient {
    public let name = "gpt"; let key: String; let model: String
    public init(key: String, model: String = "gpt-5") { self.key = key; self.model = model }
    public func complete(system: String, user: String) async throws -> String {
        var req = URLRequest(url: URL(string: "https://api.openai.com/v1/chat/completions")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.setValue("Bearer \(key)", forHTTPHeaderField: "authorization")
        req.httpBody = try JSONSerialization.data(withJSONObject: ["model": model, "messages": [["role": "system", "content": system], ["role": "user", "content": user]]])
        let (data, resp) = try await URLSession.shared.data(for: req)
        try AnthropicClient.check(resp, data)
        let j = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        return (((j["choices"] as? [[String: Any]])?.first?["message"] as? [String: Any])?["content"] as? String) ?? ""
    }
}

public struct GeminiClient: FrontierClient {
    public let name = "gemini"; let key: String; let model: String
    public init(key: String, model: String = "gemini-2.5-pro") { self.key = key; self.model = model }
    public func complete(system: String, user: String) async throws -> String {
        var req = URLRequest(url: URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent?key=\(key)")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.httpBody = try JSONSerialization.data(withJSONObject: [
            "system_instruction": ["parts": [["text": system]]],
            "contents": [["role": "user", "parts": [["text": user]]]]])
        let (data, resp) = try await URLSession.shared.data(for: req)
        try AnthropicClient.check(resp, data)
        let j = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        let parts = ((((j["candidates"] as? [[String: Any]])?.first?["content"] as? [String: Any])?["parts"] as? [[String: Any]]) ?? [])
        return parts.compactMap { $0["text"] as? String }.joined()
    }
}
